### ISSUE TITLE
EC 1+3 or standard replica?

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#592]
Let's assume I need 3 copies of each file
Which is better, an Erasure code 1+3 or a standard replication with goal 3?

EC should be able to write in parallel from clients, while standard replication write on a single chunkserver and then chunkserver replicate to each other.

EC should be faster, right ?
Do you use any hardware accelerated function for EC, like Gluster does?

**Comment by Fretek** [comment]
Hi,
you need at least 2 data parts in EC goals (LizardFS will not accept goal
ec(1, 3)), so your goal would be 2 + 3.

Yes the EC writes are parallel from clients.

The actual speed depends on configuration of the system, i.e. if you have
slower network connection between clients and chunckservers, but goo
between chunkservers then standadard goals will outperform EC.



2017-09-23 15:51 GMT+02:00 GUEST.it <notifications@github.com>:

**Comment by guestisp** [comment]
How can you have good connection between chunkservers and bad between clients and chunkservers?
AFAIK, Lizard supports only a single network for both clients and chunkservers and clients must be able to talk to all chunkservers, thus, how can you have 2 different networks ?

**Comment by Blackpaw** [comment]
I didn't realise that ec(2,3) was possible. I would have thought that if you wanted the equivalent of Replica 3 then ec(2,2) or ec(3,2) would be the way to go as that allows for up to two chunkservers going down.

You can have multiple networks/interfaces with lizardfs, just have different subnets associated with different devices, e.g

10G devices: 10.10.10.0/24

1G devices: 192.168.5.0/24

Chunkservers communicate with the master on the 10.10.10.0/24 range and mfsmount talks on the 192.168.5.0/24 range.

**Comment by 4Dolio** [comment]
I dont think you can have different interfaces on a CS for the purpose of isolating other CS vs Clients.  The CS registers with master via one address and that is the address the master tells clients to use.

**Comment by Blackpaw** [comment]
Hmmm, now I think on it .... yeah you're right :)

Clients could still be on slow (1G) interfaces and the chunkservers on fast ones though.

**Comment by 4Dolio** [comment]
Sure, link speeds need not be the same, clients could be on a 10/100 hub for that matter. We keep our masters and CS on a storage vlan/sub, and clients have a leg on that network also, to prevent crossing a router and to isolate that traffic.

**Comment by guestisp** [comment]
But this is not possible.
chunkserver are connected only with 1 IP on 1 interface.
clients must connect to these chunkserver, thus you can use a slow network connection on the client and a fast network connection on each chunkserver but this doesn't make sense, as client must talk with all chunkservers, you'll be limited by the slow client network that would be a bottleneck

maybe only chunk replications (IMHO, done directly by the chunkserver itself and not by the client like with Gluster) would be faster......

**Comment by 4Dolio** [comment]
LizardFS does not have a way to do what i think you are suggesting.  MooseFS i think does have a way to configure 'minimum write' goals where by the client only need to write X copies and the CSs will satisfy the rest via replication.  I have no first hand experience with that however and do not know if it is on the roadmap.

**Comment by guestisp** [comment]
Minimum write goal seems to be implemented some months ago

Don't know if writes are still sync up to minimum write goal and async after that or always sync but in case of chunkserver failure, writes are denied if minimum write goal is not met

**Comment by 4Dolio** [comment]
Auhh. Yes lfs has min goal which fails if unable to satisfy.

But mfs has a minimum required to be considered finished, tuned along with a timeframe i believe. Like a time delayed goal progression. Such that a new write or overwrite could be complete using a goal:mdot2+mdot2 and then at age 360 seconds changes to goal:mdot2+ssd+ssd, at age 3600 it becomes goal:ssd+hdd+hdd, at 86400 seconds it switches to goal:hdd+hdd.  All these goal changes are just more complex object properties like simple lfs goal or trashtime I believe.  They would get queued and orchestrated by the master using direct CS to CS replication mechanics.  I have not used moosefs with this feature yet but have been meaning to give it a spin...

' standard replication write on a single chunkserver and then chunkserver replicate to each other.
I think that is incorrect. normal non ec goals write occur by the client writing to all chunks copies, replication only kicks in in case of a fault in that normal write process.

Disclaimer, im still on older lfs versions mostly and dont have plans to use ec. I rarely have room for more than goal2. If i do i want all members to be capable of running independently should the need arise. Non ec is just more predictable in my situation.

**Comment by guestisp** [comment]
> I think that is incorrect. normal non ec goals write occur by the client writing to all chunks copies, replication only kicks in in case of a fault in that normal write process.

Based on official LizardFS docs, standard replication is made by chunkservers. Client writes only on a single chunkserver, than chunkserver, in turn, write to another and so on, like a chain.
https://www.researchgate.net/profile/Weigang_Wu/publication/271464202/figure/fig1/AS:295235751038978@1447401096555/Figure-1-The-readwrite-process-of-MooseFS-9.png
pic is from MooseFS, but Lizard is the same in this regard. Look at sync phase (5 and 6)

What are you describing is the GlusterFS way to replicate, where each clients write in parallel to all servers.

**Comment by 4Dolio** [comment]
I do not think that image is accurate..

**Comment by guestisp** [comment]
Only EC goals are replicated in parallel.
Standard replication is chained. It's also wrote on Lizard docs.
https://docs.lizardfs.com/adminguide/replication.html

> Note that the write modus here is: client writes chunk to ONE chunk server and this chunk server replicates this chunk to the other chunk servers.

**Comment by 4Dolio** [comment]
I saw that, after my reply i edited my post.  Now i am wondering in which version that changed...

**Comment by guestisp** [comment]
> I saw that, after my reply i edited my post. Now i am wondering in which version that changed...

I think it's in this way from the Moosefs fork, so, from the beginning.
